### DISCUSSION TITLE
Final blocks and non-blocking assignments?

### DISCUSSION THREAD

**Discussion by jrudess** [MikePopoloski/slang#405] [Q&A] [answered]
To learn the code-base, I picked (at random) a failing ivtest case to try and resolve:  https://chipsalliance.github.io/sv-tests-results/#Slang|ivtest|regress-sv_program3b_iv

To avoid Program related issues, (since no one cares about Programs :-), I changed the test case to be simpler:

```
module main;
    int unsigned test;
    final begin
        test = 3;
        test <= 5;
        if (test == 5) begin
            $display("FAILURE -- non-blocking assignment in final should not update LHS");
        end else begin
            $display("PASSED");
        end
    end
endmodule
```

The ivtest appears to want non-blocking assignments in final-blocks to be rejected at compile.  The LRM says:

> 9.2.3 Final Procedures
> 
> The only statements allowed inside a final procedure are those permitted inside a function declaration, so that they execute within a single simulation cycle.  Unlike an initial procedure, the final procedure does not execute as a separate process; instead, it executes in zero time, as a series of function calls from a single process.  All final procedures shall execute in an arbitrary order.  **No remaining scheduled events shall execute after all final procedures have executed.**

VCS accepts this code which seems LRM compliant and the ivtest appears wrong.   On the other hand, a non-blocking assignment in a final-block is nonsensical since it becomes a no-op and this is almost certainly a coding error if it occurs, so flagging an error does seem appropriate.   Any opinions?

_Note:  To implement the check, I ended up splitting out FunctionOrFinal to be independent Function/Final flags,  but there may have been a simpler method._

**Comment by MikePopoloski** [comment, marked as answer]
Hmm yeah, I think a strict reading implies this is legal, but as you say it's a useless statement. VCS allows this, Questa disallows it, and Xcelium reports a warning. I think that latter approach is best; issue an on-by-default warning that the statement is likely a mistake.

I agree that this probably means we need a separate BindFlag for Function and Final.
